### PR TITLE
linear hints

### PR DESCRIPTION
**Pull request by martyr-binbex**
Return hints in order they are processed

### CODE REVIEW

**Comment by dr-frmr** [comment]
Is there any way to retarget this to the iota-typing branch (#100 )? Logan and I refactored large parts of zink there to make jetting work.

File: lib/zink/zink.hoon [hunk 1]
```diff
@@ -1,11 +1,11 @@
 /-  *zink
 /+  *zink-pedersen, *zink-json
 =>  |%
-    +$  good      (unit *)
+    +$  good      (unit (pair * hints))
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:4]: I don't like that the return type here changed. It was factored into a `+body` as the result and an `+appendix` as all the metadata on purpose.

File: lib/zink/zink.hoon [hunk 2]
```diff
@@ -68,10 +70,12 @@
     ?~  hhed  [%&^~ app]
     =^  htal=(unit phash)  app  (hash +.f)
     ?~  htal  [%&^~ app]
-    :-  [%& ~ u.p.hed^u.p.tal]
-    app(hit [%cons u.hhed u.htal]^hit)
+    =/  hit  [%cons u.hhed u.htal]^(welp q.u.p.hed q.u.p.tal)
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:73]: unnecessary variable

File: lib/zink/zink.hoon [hunk 3]
```diff
@@ -102,41 +108,45 @@
       $(f for.f)
     ?:  ?=(%| -.formula)  ~&  103  [%|^trace app]
     ?~  p.formula  [%&^~ app]
-    %_  $
-      s    u.p.subject
-      f    u.p.formula
-      hit  [%2 u.hsub u.hfor]^hit
+    =/  res
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:111]: unnecessary variable and incorrect indentation. just keep the outer `%_` and modify the hints inline

File: lib/zink/zink.hoon [hunk 4]
```diff
@@ -178,17 +190,18 @@
       $(f subj.f)
     ?:  ?=(%| -.subject)  ~&  179  [%|^trace app]
     ?~  p.subject  [%&^~ app]
-    %_  $
-      s    u.p.subject
+    =/  res  %_  $
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:193]: same issue as before

File: lib/zink/zink.hoon [hunk 5]
```diff
@@ -197,34 +210,36 @@
       $(f head.f)
     ?:  ?=(%| -.head)  ~&  198  [%|^trace app]
     ?~  p.head  [%&^~ app]
-    %_  $
-      s    [u.p.head s]
+    =/  res  %_  $
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:213]: same issue as before

File: lib/zink/zink.hoon [hunk 6]
```diff
@@ -197,34 +210,36 @@
...
     ?~  arm  [%&^~ app]
     ?~  u.arm  ~&  216  [%|^trace app]
     =^  harm=(unit phash)  app  (hash u.u.arm)
     ?~  harm  [%&^~ app]
-    =^  hsibs=(unit (list phash))  app  (merk-sibs u.p.core axis.f)
+    =^  hsibs=(unit (list phash))  app  (merk-sibs p.u.p.core axis.f)
     ?~  hsibs  [%&^~ app]
-    %_  $
-      s    u.p.core
+    =/  res  %_  $
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:235]: same issue

File: lib/zink/zink.hoon [hunk 7]
```diff
@@ -239,31 +254,35 @@
...
     ?~  oldleaf  [%&^~ app]
     ?~  u.oldleaf  ~&  248  [%|^trace app]
     =^  holdleaf=(unit phash)  app  (hash u.u.oldleaf)
     ?~  holdleaf  [%&^~ app]
-    =^  hsibs=(unit (list phash))  app  (merk-sibs u.p.target axis.f)
+    =^  hsibs=(unit (list phash))  app  (merk-sibs p.u.p.target axis.f)
     ?~  hsibs  [%&^~ app]
-    :-  [%& ~ u.u.mutant]
-    app(hit [%10 axis.f u.hval u.htar u.holdleaf u.hsibs]^hit)
+    :-
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:268]: Line is way too long and the longer line needs to come last.

File: lib/zink/zink.hoon [hunk 8]
```diff
@@ -276,10 +295,28 @@
...
+    ~  :: Do we need hints?
   ==
+  ++  append-hints
+    |=  [[bod=body app=appendix] hin=hints]
+    ^-  book
+    ::?.  ?=([[%& ~ * bit=hints] *] b)  b
+    ?:  ?=(%| -.bod)  bod^app
+    ?~  +.bod  bod^app
+    ::=/  h  -.u.p.bod
+    [%& ~ p.u.p.bod (welp hin q.u.p.bod)]^app
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:310]: While using a function in this way is fine, I'd really like the hints to be in the appendix,

**Review by tacryt-socryp — CHANGES_REQUESTED** [review]
Please reflect the above changes and also rebase on master.

**Pull request closed without merging**